src/VisualStudio/Core/Def/LanguageService/AbstractCreateServicesOnTextViewConnection.cs

```diff
@@ -45,56 +47,54 @@ protected virtual Task InitializeServiceForOpenedDocumentAsync(Document document
+            _workQueue = new AsyncBatchingWorkQueue<DocumentId?>(
+                    TimeSpan.FromSeconds(1),
+                    ProcessBatchDocumentOpenedAsync,
+                    EqualityComparer<DocumentId?>.Default,
+                    listenerProvider.GetListener(FeatureAttribute.Workspace),
+                    threadingContext.DisposalToken);
+
+            Workspace.DocumentOpened += QueueWorkOnDocumentOpened;
         }
 
         void IWpfTextViewConnectionListener.SubjectBuffersConnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
         {
             if (!_initialized)
             {
-                var token = _listener.BeginAsyncOperation(nameof(InitializeServicesAsync));
-                InitializeServicesAsync().CompletesAsyncOperation(token);
-
                 _initialized = true;
+                // use `null` to trigger per VS session intialization task
+                _workQueue.AddWork((DocumentId?)null);
             }
         }
 
         void IWpfTextViewConnectionListener.SubjectBuffersDisconnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
         {
         }
 
-        private void InitializeServiceOnDocumentOpened(object sender, DocumentEventArgs e)
+        private async ValueTask ProcessBatchDocumentOpenedAsync(ImmutableSegmentedList<DocumentId?> documentIds, CancellationToken cancellationToken)
         {
-            if (e.Document.Project.Language != _languageName)
+            foreach (var documentId in documentIds)
             {
-                return;
-            }
-
-            var token = _listener.BeginAsyncOperation(nameof(InitializeServiceForOpenedDocumentOnBackgroundAsync));
-            InitializeServiceForOpenedDocumentOnBackgroundAsync(e.Document).CompletesAsyncOperation(token);
-
-            async Task InitializeServiceForOpenedDocumentOnBackgroundAsync(Document document)
-            {
-                await TaskScheduler.Default;
-
-                // Preload project completion providers on a background thread since loading extensions can be slow
-                // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1488945
-                if (document.GetLanguageService<CompletionService>() is not null)
-                    _ = CompletionService.GetProjectCompletionProviders(document.Project);
-
-                await InitializeServiceForOpenedDocumentAsync(document).ConfigureAwait(false);
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (documentId is null)
+                {
+                    InitializePerVSSessionServices();
+                }
+                else if (Workspace.CurrentSolution.GetDocument(documentId) is Document document && document.Project.Language == _languageName)
+                {
+                    await InitializeServiceForOpenedDocumentAsync(document).ConfigureAwait(false);
```

Comment:
(random thought unrelated to this PR, but maybe a good cleanup: the actual implementation of this method only looks at the project -- so if we're opening a bunch of files at once from the same project, we'll run it each time even though we could have just had a single batch by project ID)

Reply:
Fixed